examples/cuda/test_mm_opt.cpp

```diff
@@ -0,0 +1,401 @@
+
+
+        //************************************************************
+        // Allocate memory on the device for mat1, mat2, and out
+        //************************************************************
+        hb_mc_host_tensor_t<float> Hmat1, Hmat2, Hout, Hresult;
+
+        uint32_t M = BLOCK_DIM * 4 * dev_dim.y;
+        uint32_t N = BLOCK_DIM * 5;
+        uint32_t P = BLOCK_DIM * 3 * dev_dim.x;
+
+        eva_t _mat2, _out, _mat1;
+        hb_mc_device_tensor_t mat1, mat2, out;
+        // Set up Tensor Metadata (Host and Device)
+        // N = Number of Elements, dims = number of dimensions
+        Hmat1.N = mat1.N = M * N;
+        Hmat1.dims = mat1.dims = 2;
+
+        Hmat2.N = mat2.N = N * P;
+        Hmat2.dims = mat2.dims = 2;
+
+        Hresult.N = Hout.N = out.N = M * P;
+        Hresult.dims = Hout.dims = out.dims = 2;
+
+        // Construct mat1
+        rc = hb_mc_device_malloc(&device, sizeof(hb_mc_device_tensor_t), &_mat1);
+        rc |= hb_mc_device_malloc(&device, mat1.dims * sizeof(uint32_t), &mat1.strides);
+        rc |= hb_mc_device_malloc(&device, mat1.dims * sizeof(uint32_t), &mat1.sizes);
+        rc |= hb_mc_device_malloc(&device, mat1.N    * sizeof(*Hmat1.data), &mat1.data);
+        if (rc != HB_MC_SUCCESS) {
+                bsg_pr_err("Failed to allocate mat1 on device.\n");
+                return rc;
+        }
+
+        // Construct mat2
+        rc = hb_mc_device_malloc(&device, sizeof(hb_mc_device_tensor_t), &_mat2);
+        rc |= hb_mc_device_malloc(&device, mat2.dims * sizeof(uint32_t), &mat2.strides);
+        rc |= hb_mc_device_malloc(&device, mat2.dims * sizeof(uint32_t), &mat2.sizes);
+        rc |= hb_mc_device_malloc(&device, mat2.N    * sizeof(*Hmat2.data), &mat2.data);
+        if (rc != HB_MC_SUCCESS) {
+                bsg_pr_err("Failed to allocate mat2 on device.\n");
+                return rc;
+        }
+
+        // Construct out
+        rc = hb_mc_device_malloc(&device, sizeof(hb_mc_device_tensor_t), &_out);
+        rc |= hb_mc_device_malloc(&device, out.dims * sizeof(uint32_t), &out.strides);
+        rc |= hb_mc_device_malloc(&device, out.dims * sizeof(uint32_t), &out.sizes);
+        rc |= hb_mc_device_malloc(&device, out.N    * sizeof(*Hout.data), &out.data);
+        if (rc != HB_MC_SUCCESS) {
+                bsg_pr_err("Failed to allocate out on device.\n");
+                return rc;
+        }
+
+        // Initialize RNG
+        // Initialize the random number generators
+        std::numeric_limits<float> lim;
+
+        std::mt19937 generator;
+        generator.seed(42);
+        std::uniform_real_distribution<float> distribution(-1,1);
+
+        //************************************************************
+        // Allocate memory on the host for mat1, mat2, and result
+        // and initialize the first three with random values.
+        //************************************************************
+        Hmat1.data = new float[Hmat1.N];
```

Comment:
Probably should call `delete` on these at the end of the program (or wrap in something like an `std::unique_ptr`)